text/0026-serialization.md

```diff
@@ -0,0 +1,242 @@
+# Serialization
+
+[feature]: #feature
+
+- Start Date: `2021-07-12`
+- RFC PR: [#26](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/26)
+- SDK PR: [#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/pull/0000)
+
+# Summary
+
+[summary]: #summary
+
+This feature describes our approach to serialization. The goals of the serialization feature are that it is:
+
+- Fast
+- Allocation-free
+- Consistent
+- Easy-to-use
+- Maintainable
+
+# Motivation
+
+[motivation]: #motivation
+
+This proposes classes to enable performance improvements in MLAPI. Currently, serialization is one of our biggest offenders for CPU usage, and we've created pools to avoid allocations, which are prone to missed Dispose() calls that create leaks. This proposal is to make our serializers much faster by incorporating unsafe code operating directly on `byte*` buffers and relying on `UnsafeUtility` to tap into some native-speed operations, and replace the serializer classes with structs and NativeArrays, making it feasible to create them on the stack as needed without incurring any GC overhead.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The serialization and deserialization is done via `FastBufferWriter` and `FastBufferReader`. These are similar in interface to the current `NetworkReader` and `NetworkWriter`, having methods for serializing individual types and methods for serializing packed numbers, but in particular provide a high-performance method called `WriteValue()`/`ReadValue()` (for Writers and Readers, respectively) that can extremely quickly write an entire unmanaged struct to a buffer. There's a trade-off of CPU usage vs bandwidth in using this: Writing individual fields is slower (especially when it includes operations on unaligned memory), but allows the buffer to be filled more efficiently.
+
+Taking an example struct:
+
+```C#
+struct Message
+{
+    private float f;
+    private bool b;
+    private int i;
+
+    void Serialize(ref FastBufferWriter writer);
+}
+```
+
+Serialize can be implemented in two ways:
+
+```C#
+void Serialize(ref FastBufferWriter writer)
+{
+    writer.WriteSingle(f);
+    writer.WriteBool(b);
+    writer.WriteInt32(i);
+}
+```
+
+```C#
+void Serialize(ref FastBufferWriter writer)
+{
+    writer.WriteValue(this);
+}
+```
+
+The former will create more efficiently packed data in the message, and can be even further optimized by using `WriteSinglePacked()` and `WriteInt32Packed()`, but it has two downsides: first, that it involves more method calls and more instructions, making it slower; and second, that it creates a greater opportunity for the serialize and deserialize code to become misaligned, since they must contain the same operations in the same order.
```

Comment:
I might agree with the first part (more instructions and slower) (even though just saying slower doesn't make it undesirable immediately — how much slower? would be a better question to start thinking about better packing for bandwith vs less CPU cycles trade-off), but on the second point, I'm not sure if I understand the reason behind that argument. `Serialize()` is the method that both serializes and deserializes, there is no 2 different methods, so how could they get out of order / misordered?

Reply:
Serialize doesn't both read and write. In the IMessage proposal, Serialize writes, the MessageHandler reads. And the reason for that is that, in order to decouple message deserialization from incoming data processing (which I propose both because it simplifies maintenance by not having a massive switch statement to maintain, and because it allows us to extend IMessage functionality to user code as a better way to do custom messages), the data processing code has to not know which type it's deserializing... Ergo it can't call any instance methods. 

Also having Serialize go both ways introduces performance overhead by requiring a branch on "IsReading" in every operation... I'm trying to avoid said branches. 

As far as performance cost, it's not very high given all the methods are inlined, and I've also improved the speed of our int packing algorithm a bit. But it's not zero. I measured it a while back, but I've made some improvements for both methods since then, so I'll get some more up-to-date numbers for you.

Reply:
Ok, I have numbers for you and they widely differ between different kinds of builds.

I did these tests on a small struct:

Reply:
Tested again with something double this size. Results:

Editor:

- Test 1 took almost exactly the same amount of time it took with 4 values
- Test 2 took 210% more time vs Test 1
- Test 3 took 609% more time vs Test 1

Reply:
Also worth noting that the int values I was writing for `a` were small (just 127), so in the packed tests, they weren't even hitting the slowest path for int packing. When they do hit the slowest path, things get even worse. Setting the int and float values all to 12345678 in the double size structs I used for my last tests results in Test 3 taking **929%** more time vs Test 1.

Reply:
so, calling write multiple times is not too bad but XXXPacked() calls are 3x-9x more expensive compare to non-packed ones, right? that's not great :/

I think it's fair to comment on doing both read & write in .Serialize() wrt branching. There's a convenience vs performance dilemma there for sure. I can't really pick whether we should separate read/write or not but my gut tells me that we kind of already have/need to have both. BufferSerializer (or NetworkSerializer) might offer you .Serialize() doing both read/write based on its mode (isReading check). On the other hand BufferReader/BufferWriter could offer you a split control for serialization with their NativeArray accepting constructors operating on that array to perform read/write without isReading check. Would that be a fair design you would think?

Reply:
That is the exact design I proposed. :) BufferReader / BufferWriter are faster but harder to use, BufferSerializer is slower but more convenient and wraps Reader/Writer. My proposal is we use the faster ones internally, give our end users the convenient one, but also give them the power to pull the underlying Reader/Writer out of it if they want/need extra speed.